tests/integration/upgrade/test_sharding_rollback.py

```diff
@@ -102,36 +101,54 @@ async def test_rollback_on_config_server(
         shard_two_actual_writes == shard_two_expected_writes["number"]
     ), "continuous writes to shard two failed during upgrade"
 
+    await ops_test.model.wait_for_idle(
+        apps=CLUSTER_COMPONENTS, status="active", timeout=1000, idle_period=20
+    )
+
     # after all shards have upgraded, verify that the balancer has been turned back on
     # TODO implement this check once we have implemented the post-cluster-upgrade code DPE-4143
 
 
 @pytest.mark.runner(["self-hosted", "linux", "X64", "jammy", "large"])
 @pytest.mark.group(1)
 @pytest.mark.abort_on_fail
-@pytest.mark.skip("Need a new version published with upgrade bug fixed")
 async def test_rollback_on_shard_and_config_server(
     ops_test: OpsTest, continuous_writes_to_shard_one, continuous_writes_to_shard_two
 ) -> None:
     """Verify that a config-server and shard can safely rollback without losing writes."""
     new_charm = await ops_test.build_charm(".")
     await run_upgrade_sequence(ops_test, CONFIG_SERVER_APP_NAME, new_charm=new_charm)
 
-    shard_unit = await find_unit(ops_test, leader=True, app_name=SHARD_ONE_APP_NAME)
-    action = await shard_unit.run_action("pre-upgrade-check")
-    await action.wait()
-    assert action.status == "completed", "pre-upgrade-check failed, expected to succeed."
+    with open("charm_internal_version", mode="r") as fd:
+        revision = fd.read().strip()
+
+    # Wait for statuses to settle down
+    asyncio.gather(
+        wait_for_mongodb_units_blocked(ops_test, SHARD_ONE_APP_NAME),
+        wait_for_mongodb_units_blocked(ops_test, SHARD_TWO_APP_NAME),
+        ops_test.model.wait_for_idle(
+            apps=[CONFIG_SERVER_APP_NAME],
+            timeout=1000,
+            idle_period=20,
+            status=f"Waiting for shards to upgrade/downgrade to revision {revision}-locally built.",
+        ),
+    )
 
-    # TODO: Use this when https://github.com/juju/python-libjuju/issues/1086 is fixed
-    # await ops_test.model.applications[SHARD_ONE_APP_NAME].refresh(
-    #     channel="6/edge", switch="ch:mongodb"
-    # )
```

Comment:
This was the upgrade which is now ran as `run_upgrade_sequence`, and on top of that, the refresh here was wrong since the deployed version at this moment is already on `6/edge` so we actually want to upgrade to local charm.